commit bc4e6613f115b4e8f242ac054c9e9e708eca4031 (refs/changes/20/720/3)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2016-10-20T02:27:51+13:00 (3 years ago)
    
    PUBAPI-1310: Upgrade Node.js to v4 before 0.10 EOL in Oct 2016
    PUBAPI-1336: Add test for uncaughtexception handler

diff --git a/Makefile b/Makefile
index 753d0fa..c1e955e 100644
--- a/Makefile
+++ b/Makefile
@@ -47,9 +47,9 @@ CLEAN_FILES	+= node_modules cscope.files
 
 # The prebuilt sdcnode version we want. See
 # "tools/mk/Makefile.node_prebuilt.targ" for details.
-NODE_PREBUILT_VERSION=v0.10.47
+NODE_PREBUILT_VERSION=v4.6.1
 ifeq ($(shell uname -s),SunOS)
-	NODE_PREBUILT_IMAGE=de411e86-548d-11e4-a4b7-3bb60478632a
+	NODE_PREBUILT_IMAGE=18b094b0-eb01-11e5-80c1-175dac7ddf02
 	NODE_PREBUILT_TAG=zone
 endif
 
@@ -157,7 +157,7 @@ publish: release
 	mkdir -p $(BITS_DIR)/$(NAME)
 	cp $(ROOT)/$(RELEASE_TARBALL) $(BITS_DIR)/$(NAME)/$(RELEASE_TARBALL)
 
-.PHONY: test no_machines_test auth_test account_test analytics_test datacenters_test datasets_test fabrics_test images_test keys_test networks_test nics_test machines_all_test machines_70_test machines_71_test machines_72_test machines_73_test machines_80_test machines_test packages_test populate_networks_test services_test users_test provision_limits_plugin_test plugins_test
+.PHONY: test no_machines_test auth_test account_test analytics_test datacenters_test datasets_test fabrics_test images_test keys_test networks_test nics_test machines_all_test machines_70_test machines_71_test machines_72_test machines_73_test machines_80_test machines_test packages_test populate_networks_test services_test users_test provision_limits_plugin_test plugins_test tests_test
 
 auth_test: $(TAP)
 	$(NODE_EXEC) $(TAP) test/auth.test.js
@@ -220,12 +220,15 @@ populate_network_test: $(TAP)
 services_test: $(TAP)
 	$(NODE_EXEC) $(TAP) test/services.test.js
 
+tests_test: $(TAP)
+	$(NODE_EXEC) $(TAP) test/tests.test.js
+
 users_test: $(TAP)
 	$(NODE_EXEC) $(TAP) test/users.test.js
 
-test: auth_test account_test analytics_test datacenters_test datasets_test fabrics_test images_test keys_test networks_test packages_test populate_network_test services_test users_test nics_test machines_test
+test: auth_test account_test analytics_test datacenters_test datasets_test fabrics_test images_test keys_test networks_test packages_test populate_network_test services_test tests_test users_test nics_test machines_test
 
-no_machines_test: auth_test account_test analytics_test datacenters_test datasets_test fabrics_test images_test keys_test networks_test packages_test populate_network_test services_test users_test
+no_machines_test: auth_test account_test analytics_test datacenters_test datasets_test fabrics_test images_test keys_test networks_test packages_test populate_network_test services_test tests_test users_test
 
 provision_limits_plugin_test:
 	$(NODE_EXEC) $(TAP) test/provision_limits.test.javascript
diff --git a/docs/index.md b/docs/index.md
index e13a2fb..d497992 100644
--- a/docs/index.md
+++ b/docs/index.md
@@ -859,6 +859,15 @@ variable or the `--api-version=RANGE` option to each command.
 
 The rest of this section describes API changes in each version.
 
+
+## 8.1.0
+
+- This version should have no visible API changes, but updates a lot of
+  libraries that Cloudapi depends on, as well as the nodejs major version.
+  Visible differences with 8.0.0 are bugs, but it's possible some might have
+  crept through.
+
+
 ## 8.0.0
 
 - Instance/machine objects (from GetMachine, ListMachines) now has a `brand`
@@ -885,6 +894,7 @@ The rest of this section describes API changes in each version.
 - Network fabrics (software-defined networking) support.
   This allows the creation of virtual LANs and layer-three networks.
 
+
 ## 7.2.0
 
 - RBAC v1 has been made available on the CloudAPI interface. Accounts can create
diff --git a/etc/cloudapi.coal.cfg b/etc/cloudapi.coal.cfg
index 7b8968d..1773ada 100644
--- a/etc/cloudapi.coal.cfg
+++ b/etc/cloudapi.coal.cfg
@@ -17,7 +17,7 @@
          "*": true
     },
     "ufds": {
-        "url": "ldaps://10.99.99.18",
+        "url": "ldaps://ufds.coal.joyent.us",
         "bindDN": "cn=root",
         "bindPassword": "secret",
         "cache": {
@@ -45,31 +45,48 @@
         }
     },
     "wfapi": {
-        "url": "http://10.99.99.19"
-    },
-    "cnapi": {
-        "url": "http://10.99.99.22"
+        "url": "http://workflow.coal.joyent.us"
     },
     "vmapi": {
-        "url": "http://10.99.99.27"
+        "url": "http://vmapi.coal.joyent.us"
     },
-    "napi": {
-        "url": "http://10.99.99.10"
+    "cnapi": {
+        "url": "http://cnapi.coal.joyent.us"
     },
-    "imgapi": {
-        "url": "http://10.99.99.21"
+    "napi": {
+        "url": "http://napi.coal.joyent.us"
     },
     "fwapi": {
-        "url": "http://10.99.99.26"
+        "url": "http://fwapi.coal.joyent.us"
+    },
+    "imgapi": {
+        "url": "http://imgapi.coal.joyent.us"
     },
     "papi": {
-        "url": "http://10.99.99.29"
+        "url": "http://papi.coal.joyent.us"
     },
     "ca": {
-        "url": "http://10.99.99.30:23181"
+        "url": "http://ca.coal.joyent.us:23181"
+    },
+    "cueballHttpAgent": {
+        "resolvers": ["binder.coal.joyent.us"],
+        "initialDomains": [
+            "mahi.coal.joyent.us"
+        ],
+        "spares": 4,
+        "maximum": 10,
+        "recovery": {
+            "default": {
+                "timeout": 2000,
+                "maxTimeout": 8000,
+                "retries": 5,
+                "delay": 250,
+                "maxDelay": 1000
+            }
+        }
     },
     "mahi": {
-        "url": "http://10.99.99.33",
+        "url": "http://mahi.coal.joyent.us",
         "maxAuthCacheSize": 1000,
         "maxAuthCacheAgeMs": 300,
         "maxTranslationCacheSize": 1000,
@@ -82,7 +99,7 @@
     },
     "maxHttpSockets": 100,
     "datacenters": {
-        "coal": "http://127.0.0.1"
+        "coal": "https://cloudapi.coal.joyent.us"
     },
     "plugins": [
         {
diff --git a/lib/app.js b/lib/app.js
index 06b9ab4..d7bb7b3 100644
--- a/lib/app.js
+++ b/lib/app.js
@@ -51,6 +51,7 @@ var networks = require('./networks');
 var audit = require('./audit');
 var auditLogger = require('./audit_logger');
 var rules = require('./rules');
+var tests = require('./tests');
 
 // Account users, roles and policies:
 var users = require('./users');
@@ -85,6 +86,7 @@ var sprintf = util.format;
 
 var userThrottle = throttle.getUserThrottle;
 var ipThrottle = throttle.getIpThrottle;
+var emptyThrottle = throttle.getEmptyThrottle;
 
 var MORAY_POLL = 5000; // in ms
 
@@ -345,36 +347,7 @@ module.exports = {
         var machineThrottle;
 
         config.name = 'Joyent Triton ' + version();
-        config.version = ['8.0.0', '7.3.0', '7.2.0', '7.1.0', '7.0.0'];
-        config.formatters = {
-            'text/html': function formatHTML(req, res, body) {
-                if (typeof (body) === 'string') {
-                    return body;
-                }
-
-                var html;
-                if (body instanceof Error) {
-                    html = sprintf(HTML_FMT, body.stack);
-                } else if (Buffer.isBuffer(body)) {
-                    html = sprintf(HTML_FMT, body.toString('base64'));
-                } else {
-                    html = sprintf(HTML_FMT, body.toString());
-                }
-
-                return html;
-            },
-            'text/css': function formatCSS(req, res, body) {
-                if (typeof (body) === 'string') {
-                    return body;
-                }
-
-                return '';
-            },
-
-            'image/png': function formatPNG(req, res, body) {
-                return body;
-            }
-        };
+        config.version = ['8.1.0', '8.0.0', '7.3.0', '7.2.0', '7.1.0', '7.0.0'];
 
         if (config.dc_maint_eta) {
             var d = new Date(config.dc_maint_eta);
@@ -631,10 +604,14 @@ module.exports = {
                 policies.mount(server, userThrottle(config, 'policies'),
                         config);
                 roles.mount(server, userThrottle(config, 'roles'), config);
-                resources.mount(server, userThrottle(config, 'resources'),
-                        config);
                 nics.mount(server, userThrottle(config, 'nics'));
                 mod_config.mount(server, userThrottle(config, 'config'));
+                resources.mount(server, userThrottle(config, 'resources'),
+                        config);
+
+                if (config.test) {
+                    tests.mount(server, emptyThrottle(), config);
+                }
 
                 server.on('after', auditLogger({
                     log: log.child({component: 'audit'})
@@ -698,7 +675,7 @@ module.exports = {
                     });
                 });
 
-                server.on('uncaughtException', function (req, res, route, err) {
+                server.on('uncaughtException', function (req, res, err) {
                     var e = new restify.InternalError('Internal Error');
                     req.log.error(err, 'unexpected error');
                     res.send(e);
diff --git a/lib/machines.js b/lib/machines.js
index 2e28682..da765ea 100644
--- a/lib/machines.js
+++ b/lib/machines.js
@@ -306,11 +306,12 @@ function getListOptions(req) {
     opts.offset = req.params.offset || 0;
 
     // Copy in any and all tags
-    Object.keys(req.params).forEach(function (k) {
-        if (TAG_RE.test(k)) {
-            opts[k] = req.params[k];
-        }
-    });
+    var tags = req.params.tag;
+    if (tags) {
+        Object.keys(tags).forEach(function (tag) {
+            opts['tag.' + tag] = tags[tag];
+        });
+    }
     // By default, VMAPI will return everything unless it's told to not
     // retrieve destroyed machines.
     if (!req.params.tombstone && !opts.state) {
@@ -795,17 +796,24 @@ function xorIfPresent(a, b) {
 function loadMachine(req, res, next) {
     var pathname = req.getUrl().pathname;
 
-    if (pathname === '/--ping') {
+    if (pathname === '/--ping' || !(/\/machines/.test(pathname))) {
         return next();
     }
 
     assert.ok(req.sdc);
 
     var customer = req.account.uuid;
-    if (!req.params.machine || !(/\/machines/.test(pathname))) {
+    var name = req.params.machine;
+
+    // set by resources.js, if a PUT with role tags is being done
+    if (req.params.resource_name === 'machines' && req.params.resource_id) {
+        name = name || req.params.resource_id;
+    }
+
+    if (!name) {
         return next();
     }
-    var name = req.params.machine;
+
     var machine;
 
     return vasync.pipeline({
diff --git a/lib/tests.js b/lib/tests.js
new file mode 100644
index 0000000..4bd177e
--- /dev/null
+++ b/lib/tests.js
@@ -0,0 +1,46 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright (c) 2016, Joyent, Inc.
+ */
+
+/*
+ * This file contains functions useful during the testing of Cloudapi. They
+ * are only active when the configuration has Cloudapi in testing mode.
+ */
+
+var assert = require('assert-plus');
+
+
+function throwException(req, res, next) {
+    // non-existent function invocation should throw exception
+    req.asdasdadasdasd();
+    next();
+}
+
+
+function mount(server, before, config) {
+    assert.object(server);
+    assert.ok(before);
+    assert.ok(config);
+
+    if (!config.test) {
+        return server;
+    }
+
+    server.get({
+        path: '/:account/tests/throw_exception',
+        name: 'ThrowException'
+    }, before, throwException);
+
+    return server;
+}
+
+
+module.exports = {
+    mount: mount
+};
diff --git a/lib/throttle.js b/lib/throttle.js
index 3de867e..e15e9a6 100644
--- a/lib/throttle.js
+++ b/lib/throttle.js
@@ -60,5 +60,9 @@ module.exports = {
         assert.string(name);
 
         return getThrottle('userThrottles', config, name);
+    },
+
+    getEmptyThrottle: function () {
+        return emptyThrottle;
     }
 };
diff --git a/package.json b/package.json
index a22fb69..858efdd 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
     "name": "cloudapi",
     "description": "SmartDataCenter CloudAPI",
-    "version": "8.0.3",
+    "version": "8.1.0",
     "author": "Joyent (joyent.com)",
     "private": true,
     "engines": {
@@ -13,25 +13,25 @@
         "url": "git+ssh://git@github.com:joyent/sdc-cloudapi.git"
     },
     "dependencies": {
-        "vasync": "1.4.3",
+        "vasync": "1.6.4 ",
         "filed": "0.0.7",
         "http-signature": "1.1.0",
         "mime": "1.2.7",
-        "libuuid": "0.1.2",
+        "libuuid": "0.2.1",
         "nopt": "2.0.0",
-        "restify": "git+https://github.com/mcavage/node-restify.git#0d7b4ba",
-        "bunyan": "0.22.1",
+        "restify": "4.1.1 ",
+        "bunyan": "1.8.1",
+        "sdc-clients": "10.0.0",
         "cueball": "1.1.1",
-        "sdc-clients": "9.4.0",
-        "ufds": "1.1.3",
+        "ufds": "1.2.0",
         "semver": "2.2.1",
         "nodemailer": "0.3.29",
         "clone": "0.1.5",
         "assert-plus": "1.0.0",
         "asn1": "0.1.11",
         "ctype": "0.5.2",
-        "keyapi": "git+https://github.com/joyent/keyapi.git#c30dd2710ad2175095dc0e96479686fa774b8063",
-        "aperture": "git+https://github.com/joyent/node-aperture.git#016977",
+        "keyapi": "git+https://github.com/joyent/keyapi.git#e14b3d582e1d9d338b7082d61f34ba8d1bbc540a",
+        "aperture": "git+https://git@github.com:joyent/node-aperture.git#016977",
         "mahi": "2.0.1",
         "aperture-config": "git+https://github.com/joyent/aperture-config.git#master",
         "joyent-schemas": "git+https://github.com/joyent/schemas.git#caf3a226ed0707f5da897e1da151cc6d97fccda2",
@@ -40,7 +40,7 @@
     },
     "devDependencies": {
         "tape": "3.5.0",
-        "smartdc": "git+https://github.com/joyent/node-smartdc.git#adea68"
+        "smartdc": "8.1.0"
     },
     "sdcDependencies": {
         "imgapi": ">=2.1.0",
diff --git a/test/common.js b/test/common.js
index ac59440..9ea766a 100644
--- a/test/common.js
+++ b/test/common.js
@@ -23,7 +23,7 @@ var util = require('util');
 var fs = require('fs');
 var vasync = require('vasync');
 
-var UFDS = require('sdc-clients').UFDS;
+var UFDS = require('ufds');
 var VMAPI = require('sdc-clients').VMAPI;
 var CNAPI = require('sdc-clients').CNAPI;
 var NAPI = require('sdc-clients').NAPI;
diff --git a/test/tests.test.js b/test/tests.test.js
new file mode 100644
index 0000000..13d53e7
--- /dev/null
+++ b/test/tests.test.js
@@ -0,0 +1,64 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright (c) 2016, Joyent, Inc.
+ */
+
+var test = require('tape').test;
+var common = require('./common');
+
+
+// --- Globals
+
+
+var CLIENTS;
+var CLIENT;
+var SERVER;
+var TEST;
+
+
+// --- Tests
+
+
+test('setup', function (t) {
+    common.setup(function (_, clients, server) {
+        CLIENTS = clients;
+        CLIENT  = clients.user;
+        SERVER  = server;
+
+        TEST = common.getCfg().test;
+
+        t.end();
+    });
+});
+
+
+test('uncaughtException handler OK', function (t) {
+    if (!TEST) {
+        return t.end();
+    }
+
+    return CLIENT.get('/my/tests/throw_exception',
+            function (err, req, res, body) {
+        t.ok(err);
+        t.equal(res.statusCode, 500);
+
+        t.deepEqual(body, {
+            code: 'InternalError',
+            message: 'Internal Error'
+        });
+
+        t.end();
+    });
+});
+
+
+test('teardown', function (t) {
+    common.teardown(CLIENTS, SERVER, function () {
+        t.end();
+    });
+});
